src/providers/FileSystemProvider/FileSystemProvider.ts

```diff
@@ -184,13 +192,55 @@ export class FileSystemProvider implements vscode.FileSystemProvider {
     this._fireSoon({ type: vscode.FileChangeType.Changed, uri });
   }
 
-  public stat(uri: vscode.Uri): Promise<vscode.FileStat> {
+  public async stat(uri: vscode.Uri): Promise<vscode.FileStat> {
+    let entryPromise: Promise<Entry>;
+    let result: Entry;
     const redirectedUri = redirectDotvscodeRoot(uri);
     if (redirectedUri.path !== uri.path) {
       // When redirecting the /.vscode subtree we must fill in as-yet-unvisited folders to fix https://github.com/intersystems-community/vscode-objectscript/issues/1143
-      return this._lookup(redirectedUri, true);
+      entryPromise = this._lookup(redirectedUri, true);
+    } else {
+      entryPromise = this._lookup(uri);
+    }
+
+    // If this is our readonly variant there's no point checking server-side whether the file sould be marked readonly
+    if (uri.scheme === FILESYSTEM_READONLY_SCHEMA) {
+      return entryPromise;
+    }
+
+    if (entryPromise instanceof File) {
+      // previously resolved as a file
+      result = entryPromise;
+    } else if (entryPromise instanceof Promise && uri.path.split("/").pop()?.split(".").length > 1) {
+      // apparently a file, so resolve ahead of adding permissions
+      result = await entryPromise;
+    } else {
+      // otherwise return the promise
+      return entryPromise;
+    }
+
+    //
+    if (result instanceof File) {
+      const api = new AtelierAPI(uri);
+      const serverName = getServerName(uri);
+      if (serverName.slice(-4).toLowerCase() == ".cls") {
+        if (await isClassDeployed(serverName, api)) {
+          result.permissions |= vscode.FilePermission.Readonly;
+          return result;
+        }
+      }
+
+      // Does server-side source control report it as editable?
+      if (vscode.workspace.getConfiguration("objectscript.serverSourceControl", uri)?.get("respectEditableStatus")) {
+        const query = "select * from %Atelier_v1_Utils.Extension_GetStatus(?)";
+        const statusObj = await api.actionQuery(query, [serverName]);
+        const docStatus = statusObj.result.content.pop();
```

Comment:
```suggestion
        const docStatus = statusObj.result?.content?.pop();
```
Can't hurt to be very defensive